pkg/ingester/label_names_and_values.go

```diff
@@ -120,38 +127,46 @@ func labelValuesCardinality(
 		// For each value count total number of series storing the result into cardinality response item.
 		var respItem *client.LabelValueSeriesCount
 
-		for _, lbValue := range lbValues {
-			if err := ctx.Err(); err != nil {
-				return err
-			}
-			// Create label name response item entry.
-			if respItem == nil {
-				respItem = &client.LabelValueSeriesCount{
-					LabelName:        lbName,
-					LabelValueSeries: make(map[string]uint64),
+		resultCh := make(chan labelValueCountResult, len(lbValues))
+
+		go computeLabelValuesSeriesCount(ctx, lbName, lbValues, matchers, idxReader, postingsForMatchersFn, resultCh)
+
+		running := true
+		for running {
+			select {
+			case countRes, ok := <-resultCh:
+				if !ok {
+					running = false
+					break
 				}
-				resp.Items = append(resp.Items, respItem)
-			}
-			// Get total series count applying label matchers.
-			lblValMatchers[len(lblValMatchers)-1] = labels.MustNewMatcher(labels.MatchEqual, lbName, lbValue)
 
-			seriesCount, err := countLabelValueSeries(ctx, idxReader, postingsForMatchersFn, lblValMatchers)
-			if err != nil {
-				return err
-			}
-			respItem.LabelValueSeries[lbValue] = seriesCount
+				if countRes.err != nil {
+					return countRes.err
+				}
+				if respItem == nil {
+					respItem = &client.LabelValueSeriesCount{
+						LabelName:        lbName,
+						LabelValueSeries: make(map[string]uint64),
+					}
+					resp.Items = append(resp.Items, respItem)
+				}
+				respItem.LabelValueSeries[countRes.val] = countRes.count
 
-			respSize += len(lbValue)
-			if respSize < msgSizeThreshold {
-				continue
-			}
-			// Flush the response when reached message threshold.
-			if err := client.SendLabelValuesCardinalityResponse(srv, &resp); err != nil {
-				return err
+				respSize += len(countRes.val)
+				if respSize < msgSizeThreshold {
```

Comment:
Does there need to be one last `SendLabelValuesCardinalityResponse()` call at the end of this method?